### PR TITLE
Add show external tables

### LINKED ISSUE #2848: Implement "SHOW CREATE TABLE" for external tables
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Following on from #2529, it would be good to support external tables in "SHOW CREATE TABLE", [as spark does](https://docs.databricks.com/spark/latest/spark-sql/language-manual/sql-ref-syntax-aux-show-create-table.html).

**Describe the solution you'd like**
Implement `TableProvider::get_table_definition` for `ListingTable`, as in #2830. This will probably require implementing `Display` for `DFStatement` to convert it back into SQL, or generating SQL in `register_listing_table`.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by psvri**
# Which issue does this PR close?

Closes #2848.

 # Rationale for this change

Show create table works for external tables

# What changes are included in this PR?
Show create table now works with external tables

# Are there any user-facing changes?
Yes

### CODE REVIEW

**Comment by liukun4515** [comment]
Thanks @psvri 
Is there any SQL syntax for this function? 
It's better to add doc for this feature and it's friendly for reviewers

**Comment by psvri** [comment]
This dosent add new syntax. 
Previous create show table <external_table> was not showing the table definition. 
This PR is intended to fix that.

**Review by avantgardnerio — CHANGES_REQUESTED** [review]
Is there a reason for the syntax being `show create external table` vs just `show external table`? `Show Create` has two verbs in one sentence and seems like a confusing choice.

Also, I'm wary of passing duplicated information (the raw SQL + the result of that raw SQL) - it seems better to store the information once (in the registered table), and regenerate the SQL from that in a normalized fashion.

**Comment by avantgardnerio** [comment]
> Is there a reason for the syntax being

Oh, interesting, I see this follows the [spark precedent](https://spark.apache.org/docs/3.0.0-preview/sql-ref-syntax-aux-show-create-table.html).

**Comment by liukun4515** [comment]
> This dosent add new syntax. Previous create show table <external_table> was not showing the table definition. This PR is intended to fix that.

Can you add some SQL level test for this fix? @psvri

File: datafusion/core/tests/sql/information_schema.rs
Comment (alamb) [inline datafusion/core/tests/sql/information_schema.rs:624]: ❤️  this is great

File: datafusion/core/src/datasource/listing/table.rs
Comment (alamb) [inline datafusion/core/src/datasource/listing/table.rs:262]: This is the only thing I would prefer to change in this PR. Specifically, rather than change the API of `try_new` I think we could add a function like

```rust
pub fn with_definition(mut self, defintion: Option<String>) -> Self {
...
}
```

This change I think  would reduce the size of this PR significantly

Since I realize this PR has been waiting for a week for feedback I will propose this change via PR

**Review by alamb — COMMENTED** [review]
Sorry for the delay in reviewing @psvri  -- I have been away. We are working on increasing the review capacity for this project as well as those who can commit PRs 

All in all, I really like the feature -- I have one style suggestion

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Here is my proposed change (targeting this PR) : https://github.com/psvri/arrow-datafusion/pull/1/files

**Comment by psvri** [comment]
NP @alamb 

I too was on a break for a while. Thanks for PR. I will merge it.

**Review by alamb — APPROVED** [review]
Thanks @psvri

**Comment by alamb** [comment]
@avantgardnerio  you had an outstanding comment on this PR, I believe -- can you please review again and see if this meets your expectations now?

**Comment by avantgardnerio** [comment]
> review again and see if this meets your expectations now?

@alamb and @psvri , my concern was regarding the implementation based upon retaining the original raw SQL string. The issue this is resolving had a suggested implementation:

> This will probably require implementing Display for DFStatement to convert it back into SQL, or generating SQL in register_listing_table.

And that was more in line with my thoughts:
- we parse it into fields, and store those fields
- we also store the original parse string

This seems strange to me, in that it will carry over irrelevant syntax differences, and also stores redundant information. When I ask postgres to show a table, I don't get my raw SQL back, I get back generated SQL that matches the definition postgres has an internal representation for.

It bothers me, but I'm not sure if I'm being pedantic I'll leave it to you @alamb .

**Comment by alamb** [comment]
> This seems strange to me, in that it will carry over irrelevant syntax differences, and also stores redundant information. When I ask postgres to show a table, I don't get my raw SQL back, I get back generated SQL that matches the definition postgres has an internal representation for.

I agree that most other systems store some sort of "canonicalized" view definition.  The sqlparser-rs code offers that the output of `Display` can be parsed back into the same structure, so I think we could simply use that property (or potentially store the view as a serialized `LogicalPlan`). 

> It bothers me, but I'm not sure if I'm being pedantic I'll leave it to you @alamb .

I think the question of "how do we store a view definition" is somewhat orthogonal to the feature in this PR to show that view definition (though they are of course related!) and I think we can handle it in a follow on

Thus, I plan to file an issue where we can discus the "what to use / how to store a view definition" issue and then merge this PR. 

This may also be related to the discussion we are having about view implementation on https://github.com/apache/arrow-datafusion/issues/3249#issuecomment-1236347394

**Comment by alamb** [comment]
🤔  so I actually tried out 'show create table' with a view and it already does canonicalize the SQL, it seems:

```sql
DataFusion CLI v11.0.0
❯ create table t as select * from (values (1), (2), (3)) as sq;
+---------+
| column1 |
+---------+
| 1       |
| 2       |
| 3       |
+---------+
3 rows in set. Query took 0.050 seconds.
❯ create view v as select column1 /*my random comment*/ from t where column1 != 1;
0 rows in set. Query took 0.001 seconds.
❯ show create table v;
+---------------+--------------+------------+-----------------------------------------------------------+
| table_catalog | table_schema | table_name | definition                                                |
+---------------+--------------+------------+-----------------------------------------------------------+
| datafusion    | public       | v          | CREATE VIEW v AS SELECT column1 FROM t WHERE column1 <> 1 |
+---------------+--------------+------------+-----------------------------------------------------------+
1 row in set. Query took 0.015 seconds.
```

(note the lack of ` /*my random comment*/ ` and the transformation of `!=` to `<>`)

I believe this means that DataFusion is storing the canonicalized SQL string from sqlparser

File: datafusion/sql/src/planner.rs
Comment (alamb) [inline datafusion/sql/src/planner.rs:446]: This `to_string()` call converts the parsed statement into a string.

**Comment by alamb** [comment]
Ok, so now I am really confused. When I tried to run `show create table` locally for an external table in the datafusion-cli that didn't work 🤔 

```sql
❯ CREATE EXTERNAL TABLE abc STORED AS /*random comment*/ CSV LOCATION '../testing/data/csv/aggregate_test_100.csv';
0 rows in set. Query took 0.063 seconds.
❯ show create table abc;
+---------------+--------------+------------+------------+
| table_catalog | table_schema | table_name | definition |
+---------------+--------------+------------+------------+
| datafusion    | public       | abc        |            |
+---------------+--------------+------------+------------+
```

**Review by avantgardnerio — APPROVED** [review]
LGTM

**Comment by andygrove** [comment]
Thanks @psvri. Could you rebase or upmerge to fix the conflicts?

**Pull request merged**
